### PR TITLE
Hybrid Engine Refactor and Llama Inference Support

### PR DESCRIPTION
**Pull request by cmikeh2**
This PR introduces a number of features and bugfixes:
- The Hybrid Engine integration with Containers has been refactored. Models that support the Hybrid Engine now inherit from a feature container, either the `HybridEngineContainer` itself or something more specialized for the particular model architecture.
- Llama support for both inference and RLHF training acceleration with Hybrid Engine support
- Additional BF16 compilation support
- Additional unit test coverage for new operators and data types
- Clean up of unused code

### CODE REVIEW

File: deepspeed/module_inject/containers/base.py [hunk 1]
```diff
@@ -193,169 +191,59 @@ def mlp_quantization(self):
         self.module.mlp.inter_w = self.quantizer.quantize(self.module.mlp.inter_w)
         self.module.mlp.output_w = self.quantizer.quantize(self.module.mlp.output_w)
 
-    def apply_tensor_parallelism(self, mp_replace=None, mp_group=None, tp_size=None):
-        reversed_dim = False
-        if mp_replace is None:
-            from deepspeed.module_inject import ReplaceWithTensorSlicing
-            mp_replace = ReplaceWithTensorSlicing(mp_group=mp_group, mp_size=tp_size, out_dim=0, in_dim=1)
-            reversed_dim = True
+    def apply_tensor_parallelism(self, mp_replace=None, **kwargs):
```
Comment (awan-10) [inline deepspeed/module_inject/containers/base.py:194]: kwargs may not be needed here

File: deepspeed/module_inject/containers/base.py [hunk 2]
```diff
@@ -193,169 +191,59 @@ def mlp_quantization(self):
...
+            self.module.norm_w: self.input_nw,
+            self.module.norm_b: self.input_nb
+        }
+        for dst, src in params.items():
+            if src is None:
+                dst = src
+            else:
+                dst.data.copy_(src.to(get_accelerator().current_device_name()))
 
     def align_merged_qkv(self):
```
Comment (awan-10) [inline deepspeed/module_inject/containers/base.py:248]: remove

File: deepspeed/module_inject/containers/base.py [hunk 3]
```diff
@@ -5,11 +5,14 @@
 
 # Create a container object to save model-specific tensors using the policy file above.
 from abc import ABC
+
 import torch
 
 from deepspeed.ops.transformer.inference.config import DeepSpeedInferenceConfig
 from deepspeed.accelerator import get_accelerator
 
+DEFAULT_INTERMEDIATE_SIZE = -1
```
Comment (awan-10) [inline deepspeed/module_inject/containers/base.py:14]: please add a doc string and on the container method too

File: deepspeed/module_inject/containers/opt.py [hunk 4]
```diff
@@ -29,6 +29,32 @@ def create_module(self, config=None):
         self.module.config.scale_attention = self.scale_attention
         return self.module
 
+    def set_lora_params(self):
+        """
+        Necessry to implement for `HybridEngineContainer`
```
Comment (awan-10) [inline deepspeed/module_inject/containers/opt.py:34]: spell check

**Review by awan-10 — APPROVED** [review]

**Review by jeffra — APPROVED** [review]
🚀🚀🚀

**Review by lekurile — APPROVED** [review]

**Pull request merged**